bazel_tools/client_server_test/client_server_test.bzl

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+def _client_server_test_impl(ctx):
+
+  # Construct wrapper to execute the runner, which in turn
+  # will start the client and server.
+  wrapper = ctx.actions.declare_file(ctx.label.name + "_wrapper.sh")
+  ctx.actions.write(
+    output = wrapper,
+    content = """#!/usr/bin/env bash
+{runner} '{client}' '{server}'
+""".format(
+      runner = ctx.executable._runner.short_path,
+      client = ctx.executable.client.short_path,
+      client_args = " ".join(ctx.attr.client_args),
+      server = ctx.executable.server.short_path,
+      server_args = " ".join(ctx.attr.server_args),
+    ),
+    is_executable = True
+  )
+
+  runfiles = ctx.runfiles(files = [wrapper])
+  runfiles = runfiles.merge(ctx.attr._runner[DefaultInfo].default_runfiles)
+  runfiles = runfiles.merge(ctx.attr.client[DefaultInfo].default_runfiles)
+  runfiles = runfiles.merge(ctx.attr.server[DefaultInfo].default_runfiles)
+
+  return struct(
+    executable = wrapper,
+    files = depset([wrapper]),
+    runfiles = runfiles,
+  )
+
+client_server_test = rule(
+  implementation = _client_server_test_impl,
+  test = True,
+  executable = True,
+  attrs = {
+    "_runner": attr.label(
+      cfg = "host",
+      allow_single_file = True,
+      executable = True,
+      default = Label("@//bazel_tools/client_server_test/runner:runner"),
+    ),
+    "client": attr.label(
+      cfg = "target",
+      executable = True,
+    ),
+    "client_args": attr.string_list(),
+    "server": attr.label(
+      cfg = "target",
+      executable = True,
+    ),
+    "server_args": attr.string_list(),
+  },
+)
+"""Create a client-server test.
+
+The rule takes a client and server executables and their
+arguments as parameters. The server port is passed via a
+temporary file, which is passed to the server executable via the
+"--port-file" parameter. The client application receives the port
```

Comment:
~Out of curiousity, why is the port passed to the server in a file, and to the client as an argument?~

Got it, it's the server that writes its port into that file. The docstring is a bit ambiguous in this.

Reply:
fixed the comment.